doc/xversionmessage.md

```diff
@@ -0,0 +1,250 @@
+see https://en.bitcoin.it/wiki/Protocol_documentation#version), peers
+announce their capabilities and preferences to each other. The
+`version` message is limited in scope to a fixed set of well-known
+fields.
+
+It is deemed desirable to be able to announce further configuration
+and version information to peers. For example, this can then be used
+to version advanced block transmission protocols like Graphene.
+
+Notably, the `version` message can not be extended as there are
+clients on the current BCH network that will consider an oversized
+`version` message illegal (e.g. Bitcoin XT).
+
+For this reason, an additional message type named `xversion` is
+specified here in, plus an acknowledgement named `xverack`. The
+`xversion` message transports a generic key-value map that is meant to
+hold the configuration and version parameters.
+
+As a general reminder, it should be noted that the values given in
+this extended `xversion` version and configuration map are, except for
+basic type checks and mappings, not in any way validated and a node
+can very easily lie about them. Care should be taken that any
+implementation depending for its configuration on the values in the
+xmap will properly account for this fact and ban or otherwise penalize
+nodes that do not conform with their advertised behavior.
+
+## Encoding of the `xversion` and `xverack` message types
+
+The `xversion` message contains a single compound field which is a
+serialized key-value map henceforth named `xmap` that maps 64-bit
+integer values to variable-sized byte vectors. The message itself is
+encoded in the standard Bitcoin message frame that is documented elsewhere.
+
+The `map` is encoded using Bitcoin's usual network de-/serialization
+schemes while using `COMPACT_SIZE` size encodings whenever possible.
+Note that this is different from using the default encoding you would
+get when serializing through the `std::map` serializer (which is
+commonly found in `serialize.h` in most current C++ implementations based
+on the original Satoshi code).
+
+In particular the `xmap` payload of an `xversion` message looks like this
+for `N` entries in the map:
+```
+[compact-size N]
+[compact-size key0] [std::vector<uint8_t> default-encoding value0]
+[compact-size key1] [std::vector<uint8_t> default-encoding value1]
+...
+[compact-size key<N-1>] [std::vector<uint8_t> default-encoding value<N-1>]
+```
+
+where `std::vector<uint8_t> default-encoding` is the default encoding
+one gets when serializing a std::vector<uint8_t> of size `M` bytes through the
+serialization templates usually found in `serialize.h`:
+
+```
+[compact-size M]
+[uint8_t value0]
+...
+[uint8_t value<M-1>]
+```
+
+A node may serialize the `xmap` it wants to announce to its peers with
+any order of the key-value pairs. When receiving an `xmap`, entries
+with duplicate keys that are later in the message should override
+values that are that are closer to the start of the message.
```

Comment:
to clarify, you are suggesting using the first one as the valid one instead of the last? or that entire key becomes invalid and none of them get used?

Reply:
@dagurval: Yes, I can. Note that this mirrors the way that the default std::map deserialization works. So you'd have to pick one of your remarks.

Reply:
@awemany OK, then I agree with your choice, better to be consistent.